lib/adf_diag.py

```diff
@@ -1017,66 +1028,135 @@ def derive_variables(self, vars_to_derive=None, ts_dir=None, overwrite=None):
-                    f"ncap2 -s 'RESTOM=(FSNT-FLNT)' {constit_files[1]} {derived_file}"
-                )
+                        return None
+
+                variables = {}
+                for i in constit_list:
+                    variables[i] = ds[constit_list[0]].dims
+
+                #Get coordinate values from the first constituent file
+                coords={'lat': ds[constit_list[0]].lat.values,
+                        'lon': ds[constit_list[0]].lon.values,
+                        "time": ds[constit_list[0]].time.values}
+
+                #Create data arrays from each constituent
+                #These variables will all be added to one file that will eventually contain the
+                #derived variable as well
+                #NOTE: This has to be done via xarray dataArrays
+                #    - There might be a way with xarray dataSets but none that have worked thus far
+                data_arrays = []
+                for var_const, dims in variables.items():
+                    values = ds[var_const].values
+                    da = xr.DataArray(values, coords=coords, dims=dims)
+                    data_arrays.append(da)
+
+                #NOTE: this will need to be changed when derived equations are more than sums! - JR
+                ds[var] = sum(data_arrays)
+
+                #Aerosol Calculations - used for zonal plots
+                #These will be multiplied by rho (density of dry air)
+                ds_pmid_done = False
+                ds_t_done = False
+                if var in res["aerosol_zonal_list"]:
+
+                    #Only calculate once for all aerosol vars
+                    if not ds_pmid_done:
+                        ds_pmid = _load_dataset(glob.glob(os.path.join(ts_dir, "*.PMID.*"))[0])
+                        ds_pmid_done = True
+                        if not ds_pmid:
+                            errmsg = "Missing necessary files for rho calculation.\n"
+                            errmsg += "Please make sure 'PMID' is in the CAM run for aerosol calculations"
+                            print(errmsg)
+                            continue
+                    if not ds_t_done:
+                        ds_t = _load_dataset(glob.glob(os.path.join(ts_dir, "*.T.*"))[0])
+                        ds_t_done = True
+                        if not ds_t:
+                            errmsg = "Missing necessary files for rho calculation.\n"
+                            errmsg += "Please make sure 'T' is in the CAM run for aerosol calculations"
+                            print(errmsg)
+                            continue
+
+                    #Multiply aerosol by rho
+                    ds[var] = ds[var]*(ds_pmid["PMID"]/(res["Rgas"]*ds_t["T"]))
+                    #Sulfate conversion factor
+                    if var == "SO4":
+                        ds[var] = ds[var]*(96/115)
+
+                ds.to_netcdf(derived_file, unlimited_dims='time', mode='w')
```

Comment:
I think that ds has the constituents and the derived values? 

Do we want to store all of them in the output file? I think in the version using ncap2 that was necessary because of how NCO works, but since we already have the constituents in files, I think this would be replicating data.